# DATA CLEANING + EDA (Pandas cont.)

Created By: Angelica Rojas

In [581]:
import pandas as pd
import re

## Upload Data

The data for this notebook could be found at this link: https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5 . 

For the purpose of this lesson we will use the CSV file of the data.

In [582]:
df = pd.read_csv("BerkeleyPD_Calls_for_Service.csv")

#what does this do?
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

             CVLEGEND  CVDOW                InDbDate  \
0  BURGLARY - VEHICLE      3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE      5  09/25/2017 03:30:18 AM   
2             LARCENY      6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE      0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION      6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

Why did we only want to display the first 5 rows of the dataframe?

What if we wanted to see the size of this dataframe?

In [583]:
# number of rows
len(df.index)

5617

In [584]:
# shape of df (rows, columns)

df.shape

(5617, 11)

# Part 1: DATA CLEANING

## Column Names

What do all these column names even mean? 

On that same website, BPD offers a narrative pdf file that describes the data they provided. https://data.cityofberkeley.info/Public-Safety/Berkeley-PD-Calls-for-Service/k2nh-s5h5


<img src = "DF_col_desc.png">

## Change Column Name(s)

Why would we want to change the column names?

In [585]:
#df = df.rename(columns={'ORIG_COL_NAME': 'NEW_COL_NAME'})
df = df.rename(columns={'CVLEGEND': 'EVENTDESC', 'CVDOW':"D.O.W."})
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  
2               2500 SHATTUCK AVE  Berkeley    CA  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA

In [586]:
df.EVENTDT.min()

'03/28/2017 12:00:00 AM'

## Investigating Columns

What is the difference between "Block_Location" and "BLKADDR" columns in the dataframe? From the look of the DF that is displayed it looks as though they are almost the same.

In [587]:
#Let's look at the first value in "Block_Location"
df["Block_Location"][0]

'ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n(37.869363, -122.268028)'

In [588]:
#Let's look at the first value in "BLKADDR"
df["BLKADDR"][0]

'ALLSTON WAY & SHATTUCK AVE'

## Create New Columns

What is new information about "Block_Location" that we can actually use and save?

Let's create new columns for the information we extracted from those values.

In [589]:
coordinates = [["".join(x.split()) for x in re.split(r'[()]',i) if x.strip()][-1] for i in df["Block_Location"]]

#new values
longitude =[["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][-1] for i in coordinates]
latitude = [["".join(x.split()) for x in re.split(r'[,]',i) if x.strip()][0] for i in coordinates]

#create new columns for latitude and longitude
df["LATITUDE"] = latitude
df["LONGITUDE"] = longitude

#Check if it worked
df.head()



CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                                      Block_Location  \
0  ALLSTON WAY &amp; SHATTUCK AVE\nBerkeley, CA\n...   
1  UNIVERSITY AVENUE &amp; MILVIA ST\nBerkeley, C...   
2  2500 SHATTUCK AVE\nBerkeley, CA\n(37.863811, -...   
3  DURANT AVENUE &amp; ELLSWORTH ST\nBerkeley, CA...   
4  MILVIA STREET &amp; UNIVERSITY AVE\nBerkeley, ...   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

In [590]:
df.LATITUDE.unique()

array(['37.869363', '37.871884', '37.863811', ..., '37.877247',
       '37.874537', '37.861541'], dtype=object)

In [591]:
df.LATITUDE
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5617 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5617 non-null int64
OFFENSE           5617 non-null object
EVENTDT           5617 non-null object
EVENTTM           5617 non-null object
EVENTDESC         5617 non-null object
D.O.W.            5617 non-null int64
InDbDate          5617 non-null object
Block_Location    5617 non-null object
BLKADDR           5590 non-null object
City              5617 non-null object
State             5617 non-null object
LATITUDE          5617 non-null object
LONGITUDE         5617 non-null object
dtypes: int64(2), object(11)
memory usage: 570.5+ KB


In [592]:
df2 = df[df['LATITUDE'].str.startswith('37')]

In [593]:
df2.LATITUDE

0       37.869363
1       37.871884
2       37.863811
3       37.867221
4       37.871884
5       37.868706
6       37.865849
7       37.880266
8       37.882457
9       37.856195
10      37.879708
11      37.859557
12      37.868352
13      37.821533
14      37.892152
15      37.878644
16       37.87091
17      37.887498
18      37.861672
19      37.858214
20      37.853275
22       37.86939
23      37.858165
24      37.859557
25       37.88015
26      37.880262
27      37.858628
28      37.880227
29      37.867852
30      37.861604
          ...    
5585    37.855435
5586    37.881141
5587    37.874489
5588     37.87247
5589    37.857409
5590    37.863072
5591    37.866426
5592    37.875531
5593     37.84827
5594      37.8704
5595    37.867212
5596    37.863839
5597    37.870417
5598    37.869921
5599    37.857099
5600    37.865711
5601    37.870289
5602    37.888784
5603    37.864826
5604    37.864535
5605    37.857674
5607    37.868512
5608    37.859309
5610    37.865816
5611    37

In [594]:
df.LONGITUDE.unique()

array(['-122.268028', '-122.270752', '-122.267412', ..., '-122.27708',
       '-122.263862', '-122.251156'], dtype=object)

In [595]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5617 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5617 non-null int64
OFFENSE           5617 non-null object
EVENTDT           5617 non-null object
EVENTTM           5617 non-null object
EVENTDESC         5617 non-null object
D.O.W.            5617 non-null int64
InDbDate          5617 non-null object
Block_Location    5617 non-null object
BLKADDR           5590 non-null object
City              5617 non-null object
State             5617 non-null object
LATITUDE          5617 non-null object
LONGITUDE         5617 non-null object
dtypes: int64(2), object(11)
memory usage: 570.5+ KB


In [596]:
df2 = df[df['LONGITUDE'].str.startswith('-122')]
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5337 entries, 0 to 5616
Data columns (total 13 columns):
CASENO            5337 non-null int64
OFFENSE           5337 non-null object
EVENTDT           5337 non-null object
EVENTTM           5337 non-null object
EVENTDESC         5337 non-null object
D.O.W.            5337 non-null int64
InDbDate          5337 non-null object
Block_Location    5337 non-null object
BLKADDR           5310 non-null object
City              5337 non-null object
State             5337 non-null object
LATITUDE          5337 non-null object
LONGITUDE         5337 non-null object
dtypes: int64(2), object(11)
memory usage: 583.7+ KB


In [597]:
df3 = df[df['LONGITUDE'].str.startswith('-122') & df['LATITUDE'].str.startswith('37')]


In [598]:
df3['LATITUDE'].nunique()

1383

In [599]:
len(df3[['LONGITUDE', 'LATITUDE']].index)

5337

## Drop Columns

We got all the information we needed from "Block_Location" keeping it would be taking up extra room in our dataframe. 

Let's drop the "Block_Location" from the dataframe.

In [600]:
#df = df.drop("COL_NAME", axis = 1)
df = df.drop("Block_Location", axis = 1)
#Check if it dropped
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                          BLKADDR      City State   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  Berkeley    CA  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  Berkeley    CA  37.871884  -122.270752  
2               2500 SHATTUCK AVE  Berkeley    CA  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  Berkeley    CA  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  Berkeley    CA  37.871884  -122.270752

We can drop other columns that we do not think would add useful information to our analysis. 

Although we did get this data from the Berkeley PD, let's make sure all values in "City" are "Berkeley". Also, Let's make sure the "State" is "CA" for all values.

In [601]:
df.City.unique()

array(['Berkeley'], dtype=object)

In [602]:
df['State'].unique()

array(['CA'], dtype=object)

We checked all the unique values for columns "State" and "City" and they are the results we wanted, therefore, we do not need those columns anymore. 

Drop the columns listed above.

In [603]:
#drop City and State columns
#df = ...
df = df.drop(["City", "State"], axis = 1)
#Check if they dropped
df.head()

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
0  17034187             BURGLARY AUTO  06/14/2017 12:00:00 AM   15:15   
1  17052235                GUN/WEAPON  09/01/2017 12:00:00 AM   22:56   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017 12:00:00 AM   10:45   
3  17018444             BURGLARY AUTO  04/02/2017 12:00:00 AM   19:30   
4  17033328                 NARCOTICS  06/10/2017 12:00:00 AM   14:30   

            EVENTDESC  D.O.W.                InDbDate  \
0  BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM   
1     WEAPONS OFFENSE       5  09/25/2017 03:30:18 AM   
2             LARCENY       6  09/25/2017 03:30:15 AM   
3  BURGLARY - VEHICLE       0  09/25/2017 03:30:11 AM   
4      DRUG VIOLATION       6  09/25/2017 03:30:14 AM   

                          BLKADDR   LATITUDE    LONGITUDE  
0      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2               2500 SHATTUCK AVE  37.863811  -122.267412  
3    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

## Dealing With Null Values (NaN)

<img src = "null_def.png">

REFERENCE: https://pandas.pydata.org/pandas-docs/stable/missing_data.html

This is a big data set and we can't look through each value one at a time. How can we make sure that there is a value for each category?

In [604]:
df.isnull().sum()

CASENO        0
OFFENSE       0
EVENTDT       0
EVENTTM       0
EVENTDESC     0
D.O.W.        0
InDbDate      0
BLKADDR      27
LATITUDE      0
LONGITUDE     0
dtype: int64

Let's look at the rows where "BLKADDR" is a null value. Let's make a temporary sub-dataframe.

In [605]:
null_temp = df[pd.isnull(df['BLKADDR'])]
null_temp

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
53    17036936               DISTURBANCE  06/26/2017 12:00:00 AM   18:24   
104   17090713  THEFT FELONY (OVER $950)  04/09/2017 12:00:00 AM   04:15   
224   17024641             BURGLARY AUTO  05/01/2017 12:00:00 AM   21:00   
235   17046547            VEHICLE STOLEN  08/08/2017 12:00:00 AM   17:00   
291   17053694  THEFT MISD. (UNDER $950)  09/07/2017 12:00:00 AM   17:43   
475   17022572            VEHICLE STOLEN  04/22/2017 12:00:00 AM   21:00   
534   17026854      BURGLARY RESIDENTIAL  05/12/2017 12:00:00 AM   09:00   
1228  17091147             BURGLARY AUTO  06/14/2017 12:00:00 AM   03:00   
1306  17020446            VEHICLE STOLEN  04/12/2017 12:00:00 AM   18:00   
1311  17025351           THEFT FROM AUTO  05/04/2017 12:00:00 AM   22:30   
1550  17029419      BURGLARY RESIDENTIAL  05/23/2017 12:00:00 AM   16:30   
1624  17033574                 NARCOTICS  06/11/2017 12:00:00 AM   16:53   
2093  17037072             BURGLARY AUTO  06/20/2017 12:00:00 AM   08:00   
2340  17028899           ALCOHOL OFFENSE  05/21/2017 12:00:00 AM   22:00   
2595  17030776      BURGLARY RESIDENTIAL  05/30/2017 12:00:00 AM   10:45   
3453  17038382      BURGLARY RESIDENTIAL  06/30/2017 12:00:00 AM   15:00   
3922  17034863               DISTURBANCE  06/01/2017 12:00:00 AM   00:00   
4074  17091632                 VANDALISM  08/15/2017 12:00:00 AM   23:30   
4505  17091900  THEFT MISD. (UNDER $950)  09/20/2017 12:00:00 AM   21:35   
4545  17026051           THEFT FROM AUTO  05/07/2017 12:00:00 AM   16:00   
4759  17091711             BURGLARY AUTO  08/28/2017 12:00:00 AM   08:20   
4852  17042511             BURGLARY AUTO  07/20/2017 12:00:00 AM   16:00   
4959  17090810             BURGLARY AUTO  04/23/2017 12:00:00 AM   00:00   
5032  17030164            VEHICLE STOLEN  05/26/2017 12:00:00 AM   20:00   
5194  17091379             BURGLARY AUTO  07/14/2017 12:00:00 AM   12:00   
5406  17020088     ASSAULT/BATTERY MISD.  04/11/2017 12:00:00 AM   09:25   
5544  17091287                 VANDALISM  07/01/2017 12:00:00 AM   08:00   

                   EVENTDESC  D.O.W.                InDbDate BLKADDR  \
53        DISORDERLY CONDUCT       1  09/25/2017 03:30:16 AM     NaN   
104                  LARCENY       0  09/25/2017 03:30:12 AM     NaN   
224       BURGLARY - VEHICLE       1  09/25/2017 03:30:12 AM     NaN   
235      MOTOR VEHICLE THEFT       2  09/25/2017 03:30:18 AM     NaN   
291                  LARCENY       4  09/25/2017 03:30:19 AM     NaN   
475      MOTOR VEHICLE THEFT       6  09/25/2017 03:30:12 AM     NaN   
534   BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:12 AM     NaN   
1228      BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM     NaN   
1306     MOTOR VEHICLE THEFT       3  09/25/2017 03:30:12 AM     NaN   
1311  LARCENY - FROM VEHICLE       4  09/25/2017 03:30:12 AM     NaN   
1550  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:13 AM     NaN   
1624          DRUG VIOLATION       0  09/25/2017 03:30:14 AM     NaN   
2093      BURGLARY - VEHICLE       2  09/25/2017 03:30:16 AM     NaN   
2340    LIQUOR LAW VIOLATION       0  09/25/2017 03:30:13 AM     NaN   
2595  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:14 AM     NaN   
3453  BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:17 AM     NaN   
3922      DISORDERLY CONDUCT       4  09/25/2017 03:30:16 AM     NaN   
4074               VANDALISM       2  09/25/2017 03:30:18 AM     NaN   
4505                 LARCENY       3  09/25/2017 03:30:19 AM     NaN   
4545  LARCENY - FROM VEHICLE       0  09/25/2017 03:30:12 AM     NaN   
4759      BURGLARY - VEHICLE       1  09/25/2017 03:30:18 AM     NaN   
4852      BURGLARY - VEHICLE       4  09/25/2017 03:30:17 AM     NaN   
4959      BURGLARY - VEHICLE       0  09/25/2017 03:30:12 AM     NaN   
5032     MOTOR VEHICLE THEFT       5  09/25/2017 03:30:13 AM     NaN   
5194      BURGLARY - VEHICLE       5  09/25/2017 03:30:17 AM     NaN   
5406   

Does the number of rows in the dataframe match the values above?

In [606]:
#get number of rows of new df
len(null_temp)


27

Investigate the dataframe, do you see somethng interesting that all these rows share?

Are the Latitude/Longitude values all the same for the "NaN" values?

In [607]:
#get unique values of latitude

null_temp.LATITUDE.unique()

array(['37.869058'], dtype=object)

In [608]:
#get unique values of longitude

null_temp.LONGITUDE.unique()

array(['-122.270455'], dtype=object)

## Boolean Slicing

Let's look at the whole dataset to see if there are any rows with that Latitude and Longitude combination that might have a "BLKADDR" associated with it.  

In [609]:
df[(df["LATITUDE"] == '37.869058') & (df["LONGITUDE"] == '-122.270455')]

CASENO                   OFFENSE                 EVENTDT EVENTTM  \
53    17036936               DISTURBANCE  06/26/2017 12:00:00 AM   18:24   
104   17090713  THEFT FELONY (OVER $950)  04/09/2017 12:00:00 AM   04:15   
224   17024641             BURGLARY AUTO  05/01/2017 12:00:00 AM   21:00   
235   17046547            VEHICLE STOLEN  08/08/2017 12:00:00 AM   17:00   
291   17053694  THEFT MISD. (UNDER $950)  09/07/2017 12:00:00 AM   17:43   
475   17022572            VEHICLE STOLEN  04/22/2017 12:00:00 AM   21:00   
534   17026854      BURGLARY RESIDENTIAL  05/12/2017 12:00:00 AM   09:00   
1228  17091147             BURGLARY AUTO  06/14/2017 12:00:00 AM   03:00   
1306  17020446            VEHICLE STOLEN  04/12/2017 12:00:00 AM   18:00   
1311  17025351           THEFT FROM AUTO  05/04/2017 12:00:00 AM   22:30   
1550  17029419      BURGLARY RESIDENTIAL  05/23/2017 12:00:00 AM   16:30   
1624  17033574                 NARCOTICS  06/11/2017 12:00:00 AM   16:53   
2093  17037072             BURGLARY AUTO  06/20/2017 12:00:00 AM   08:00   
2340  17028899           ALCOHOL OFFENSE  05/21/2017 12:00:00 AM   22:00   
2595  17030776      BURGLARY RESIDENTIAL  05/30/2017 12:00:00 AM   10:45   
3453  17038382      BURGLARY RESIDENTIAL  06/30/2017 12:00:00 AM   15:00   
3922  17034863               DISTURBANCE  06/01/2017 12:00:00 AM   00:00   
4074  17091632                 VANDALISM  08/15/2017 12:00:00 AM   23:30   
4505  17091900  THEFT MISD. (UNDER $950)  09/20/2017 12:00:00 AM   21:35   
4545  17026051           THEFT FROM AUTO  05/07/2017 12:00:00 AM   16:00   
4759  17091711             BURGLARY AUTO  08/28/2017 12:00:00 AM   08:20   
4852  17042511             BURGLARY AUTO  07/20/2017 12:00:00 AM   16:00   
4959  17090810             BURGLARY AUTO  04/23/2017 12:00:00 AM   00:00   
5032  17030164            VEHICLE STOLEN  05/26/2017 12:00:00 AM   20:00   
5194  17091379             BURGLARY AUTO  07/14/2017 12:00:00 AM   12:00   
5406  17020088     ASSAULT/BATTERY MISD.  04/11/2017 12:00:00 AM   09:25   
5544  17091287                 VANDALISM  07/01/2017 12:00:00 AM   08:00   

                   EVENTDESC  D.O.W.                InDbDate BLKADDR  \
53        DISORDERLY CONDUCT       1  09/25/2017 03:30:16 AM     NaN   
104                  LARCENY       0  09/25/2017 03:30:12 AM     NaN   
224       BURGLARY - VEHICLE       1  09/25/2017 03:30:12 AM     NaN   
235      MOTOR VEHICLE THEFT       2  09/25/2017 03:30:18 AM     NaN   
291                  LARCENY       4  09/25/2017 03:30:19 AM     NaN   
475      MOTOR VEHICLE THEFT       6  09/25/2017 03:30:12 AM     NaN   
534   BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:12 AM     NaN   
1228      BURGLARY - VEHICLE       3  09/25/2017 03:30:15 AM     NaN   
1306     MOTOR VEHICLE THEFT       3  09/25/2017 03:30:12 AM     NaN   
1311  LARCENY - FROM VEHICLE       4  09/25/2017 03:30:12 AM     NaN   
1550  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:13 AM     NaN   
1624          DRUG VIOLATION       0  09/25/2017 03:30:14 AM     NaN   
2093      BURGLARY - VEHICLE       2  09/25/2017 03:30:16 AM     NaN   
2340    LIQUOR LAW VIOLATION       0  09/25/2017 03:30:13 AM     NaN   
2595  BURGLARY - RESIDENTIAL       2  09/25/2017 03:30:14 AM     NaN   
3453  BURGLARY - RESIDENTIAL       5  09/25/2017 03:30:17 AM     NaN   
3922      DISORDERLY CONDUCT       4  09/25/2017 03:30:16 AM     NaN   
4074               VANDALISM       2  09/25/2017 03:30:18 AM     NaN   
4505                 LARCENY       3  09/25/2017 03:30:19 AM     NaN   
4545  LARCENY - FROM VEHICLE       0  09/25/2017 03:30:12 AM     NaN   
4759      BURGLARY - VEHICLE       1  09/25/2017 03:30:18 AM     NaN   
4852      BURGLARY - VEHICLE       4  09/25/2017 03:30:17 AM     NaN   
4959      BURGLARY - VEHICLE       0  09/25/2017 03:30:12 AM     NaN   
5032     MOTOR VEHICLE THEFT       5  09/25/2017 03:30:13 AM     NaN   
5194      BURGLARY - VEHICLE       5  09/25/2017 03:30:17 AM     NaN   
5406   

In [610]:
#get unique values of BLKADDR for the Lat/Long combo



## Drop Null Values (NaN)

We could essentially go to Google and try to figure out the BLKADDR ourselves, but to remove any problems that may occur while searching let's just drop all the rows that include null values

In [611]:
#drop rows that have null values
df = df.dropna(axis = 0, how = "any")

In [612]:
#now find out how many null values within the df
#What should you see when you run this?


# <font color = "red"> YOUR TURN! </font>

### What is the difference between "EVENTDT" and "EVENTTM"? How can we clean our columns to reflect the data that is useful?

HINT: Focus on EVENTDT

In [613]:
#Slice the string to get the information you want and set to the variable
date = [i[:10] for i in df["EVENTDT"]]

#Replace "EVENTDT" with new variable
df["EVENTDT"] = date
df.head()
#check if it worked
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

   D.O.W.                InDbDate                         BLKADDR   LATITUDE  \
0       3  09/25/2017 03:30:15 AM      ALLSTON WAY & SHATTUCK AVE  37.869363   
1       5  09/25/2017 03:30:18 AM   UNIVERSITY AVENUE & MILVIA ST  37.871884   
2       6  09/25/2017 03:30:15 AM               2500 SHATTUCK AVE  37.863811   
3       0  09/25/2017 03:30:11 AM    DURANT AVENUE & ELLSWORTH ST  37.867221   
4       6  09/25/2017 03:30:14 AM  MILVIA STREET & UNIVERSITY AVE  37.871884   

     LONGITUDE  
0  -122.268028  
1  -122.270752  
2  -122.267412  
3  -122.263531  
4  -122.270752

### Drop the "InDbDate" column

In [614]:
df = df.drop("InDbDate", axis = 1)

#Check that it actually dropped
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

   D.O.W.                         BLKADDR   LATITUDE    LONGITUDE  
0       3      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1       5   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2       6               2500 SHATTUCK AVE  37.863811  -122.267412  
3       0    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4       6  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

### TRICKY QUESTION

### Column "D.O.W." can be a bit confusing with the numbers. Replace the numbers with the appropriate day it corresponds to. 

You can find the days it corresponds to in the beginning of the notebook. 

##### HINT: You may need to use a dictionary, the map function, or the zip function

DICT:
https://www.programiz.com/python-programming/methods/built-in/dict

MAP:
https://www.programiz.com/python-programming/methods/built-in/map

ZIP:
https://www.programiz.com/python-programming/methods/built-in/zip

In [615]:
dow = {0:"Sunday",1:"Monday",2:"Tuesday",3:"Wednesday", 4:"Thursday", 5:"Friday", 6:"Saturday"}
df["D.O.W."] = df["D.O.W."].map(dow)

#Check if it worked
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                         BLKADDR   LATITUDE    LONGITUDE  
0  Wednesday      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1     Friday   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2   Saturday               2500 SHATTUCK AVE  37.863811  -122.267412  
3     Sunday    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4   Saturday  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

# Part 2: EXPLORATORY DATA ANALYSIS

<h3>"Exploratory data analysis or 'EDA' is a <b>critical</b> beginning step in analyzing the data from an experiment.</h3>

<b>Here are the main reasons we use EDA:</b>
<ul>
• detection of mistakes<br><br>
• checking of assumptions<br><br>
• preliminary selection of appropriate models<br><br>
• determining relationships among the explanatory variables, and<br><br>
• assessing the direction and rough size of relationships between explanatory and outcome variables."</ul>
REFERENCE: http://www.stat.cmu.edu/~hseltman/309/Book/chapter4.pdf


## Now what?

We have cleaned our data to the best of our ability based on the initial look. Now let's try to look at the <b>relationships</b> between different values. 

In [616]:
df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                         BLKADDR   LATITUDE    LONGITUDE  
0  Wednesday      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1     Friday   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2   Saturday               2500 SHATTUCK AVE  37.863811  -122.267412  
3     Sunday    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4   Saturday  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

Let's look at the different types of offenses that were called in. We know that using the .unique() function will return all the unique values in the column, but what if we wanted to also <b>count</b> the different times each unique value appeared?

In [617]:
df.OFFENSE.value_counts()

BURGLARY AUTO               1069
THEFT MISD. (UNDER $950)     864
VANDALISM                    447
DISTURBANCE                  423
NARCOTICS                    339
VEHICLE STOLEN               306
BURGLARY RESIDENTIAL         271
ASSAULT/BATTERY MISD.        263
THEFT FELONY (OVER $950)     253
ROBBERY                      183
IDENTITY THEFT               150
ALCOHOL OFFENSE              141
THEFT FROM AUTO              137
DOMESTIC VIOLENCE            119
BURGLARY COMMERCIAL          111
ASSAULT/BATTERY FEL.         102
FRAUD/FORGERY                101
MISSING ADULT                 66
2ND RESPONSE                  42
GUN/WEAPON                    42
SEXUAL ASSAULT FEL.           32
BRANDISHING                   25
THEFT FROM PERSON             23
MISSING JUVENILE              22
SEXUAL ASSAULT MISD.          21
ARSON                         16
MUNICIPAL CODE                12
VEHICLE RECOVERED              8
KIDNAPPING                     1
VICE                           1
Name: OFFE

In [618]:
df.EVENTDESC.value_counts()

LARCENY                   1140
BURGLARY - VEHICLE        1069
VANDALISM                  447
DISORDERLY CONDUCT         424
ASSAULT                    365
DRUG VIOLATION             339
MOTOR VEHICLE THEFT        306
BURGLARY - RESIDENTIAL     271
FRAUD                      251
ROBBERY                    183
LIQUOR LAW VIOLATION       141
LARCENY - FROM VEHICLE     137
FAMILY OFFENSE             119
BURGLARY - COMMERCIAL      111
MISSING PERSON              88
WEAPONS OFFENSE             67
SEX CRIME                   53
NOISE VIOLATION             42
ARSON                       16
ALL OTHER OFFENSES          12
RECOVERED VEHICLE            8
KIDNAPPING                   1
Name: EVENTDESC, dtype: int64

Why is "LARCENY" a higher occurence in the "EVENTDESC" column, if when we looked into the "OFFENSE" column, "BURGLARY - VEHICLE" is first? Let's look into this a little more.


## GroupBy 

In [619]:
df.groupby("EVENTDESC").OFFENSE.value_counts()

#turn the series into a DF 
df.groupby("EVENTDESC").OFFENSE.value_counts().to_frame()

OFFENSE
EVENTDESC              OFFENSE                          
ALL OTHER OFFENSES     MUNICIPAL CODE                 12
ARSON                  ARSON                          16
ASSAULT                ASSAULT/BATTERY MISD.         263
                       ASSAULT/BATTERY FEL.          102
BURGLARY - COMMERCIAL  BURGLARY COMMERCIAL           111
BURGLARY - RESIDENTIAL BURGLARY RESIDENTIAL          271
BURGLARY - VEHICLE     BURGLARY AUTO                1069
DISORDERLY CONDUCT     DISTURBANCE                   423
                       VICE                            1
DRUG VIOLATION         NARCOTICS                     339
FAMILY OFFENSE         DOMESTIC VIOLENCE             119
FRAUD                  IDENTITY THEFT                150
                       FRAUD/FORGERY                 101
KIDNAPPING             KIDNAPPING                      1
LARCENY                THEFT MISD. (UNDER $950)      864
                       THEFT FELONY (OVER $950)      253
                       THEFT FROM PERSON              23
LARCENY - FROM VEHICLE THEFT FROM AUTO               137
LIQUOR LAW VIOLATION   ALCOHOL OFFENSE               141
MISSING PERSON         MISSING ADULT                  66
                       MISSING JUVENILE               22
MOTOR VEHICLE THEFT    VEHICLE STOLEN                306
NOISE VIOLATION        2ND RESPONSE                   42
RECOVERED VEHICLE      VEHICLE RECOVERED               8
ROBBERY                ROBBERY                       183
SEX CRIME              SEXUAL ASSAULT FEL.            32
                       SEXUAL ASSAULT MISD.           21
VANDALISM              VANDALISM                     447
WEAPONS OFFENSE        GUN/WEAPON                     42
                       BRANDISHING                    25

From this DF we can see that "LARCENY" has the most OFFENSES within that category. When you add the totals from "THEFT MISD ( UNDER $ 950)" ," THEFT  FELONY (OVER $ 950)", and
"THEFT FROM PERSON" they equal more than "BURGLARY - VEHICLE", but "BURGLARY AUTO" as an offense alone is the highest in number.

# <font color = "red"> YOUR TURN! </font>

Could there be any relationship with the Day of the Week and the calls? Try out different functions to see if there is any significance?

In [620]:
#count the amount of calls per day
df["D.O.W."].value_counts()

Tuesday      853
Friday       818
Wednesday    818
Saturday     809
Thursday     797
Monday       793
Sunday       702
Name: D.O.W., dtype: int64

With the day that has the most calls, check the type of offense that appears the most.

In [621]:
#only display rows with the D.O.W that appears the most
#create a temp df


In [622]:
#only display rows with the D.O.W of Friday
#create a temp df

fri =df[df["D.O.W."] == "Friday"]
fri.head()

CASENO                   OFFENSE     EVENTDT EVENTTM  \
1   17052235                GUN/WEAPON  09/01/2017   22:56   
5   17043981           ALCOHOL OFFENSE  07/28/2017   21:31   
6   17091423  THEFT FELONY (OVER $950)  07/21/2017   20:35   
14  17090740            IDENTITY THEFT  04/14/2017   12:55   
16  17048522     ASSAULT/BATTERY MISD.  08/18/2017   07:45   

               EVENTDESC  D.O.W.                        BLKADDR   LATITUDE  \
1        WEAPONS OFFENSE  Friday  UNIVERSITY AVENUE & MILVIA ST  37.871884   
5   LIQUOR LAW VIOLATION  Friday   KITTREDGE STREET & FULTON ST  37.868706   
6                LARCENY  Friday                  2400 HASTE ST  37.865849   
14                 FRAUD  Friday              900 SAN BENITO RD  37.892152   
16               ASSAULT  Friday                 2100 OXFORD ST   37.87091   

      LONGITUDE  
1   -122.270752  
5   -122.266279  
6   -122.259977  
14  -122.269211  
16  -122.265993

Do these numbers match the results of the overall DF?


Let's try something else, Friday and Saturday nights are typically associated with being the "party" time. If this is true should there be more Liquor/Drug/Disordely Conduct/etc. occurrences those nights?

Let's try it with Fridays!

This will not give us the information we want. Instead let us look at each EVENTDESC and group by the D.O.W. that appears the most per EVENTDESC.

Also, I realized that "D.O.W." is becoming a problem with the periods when I am trying to call my series. I want to change the name of the column again, how can I do that?

In [623]:
#change column name

df.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                         BLKADDR   LATITUDE    LONGITUDE  
0  Wednesday      ALLSTON WAY & SHATTUCK AVE  37.869363  -122.268028  
1     Friday   UNIVERSITY AVENUE & MILVIA ST  37.871884  -122.270752  
2   Saturday               2500 SHATTUCK AVE  37.863811  -122.267412  
3     Sunday    DURANT AVENUE & ELLSWORTH ST  37.867221  -122.263531  
4   Saturday  MILVIA STREET & UNIVERSITY AVE  37.871884  -122.270752

# <font color = "red"> GROUP WORK</font> 
## What do YOU want to find out? YOUR DATA INVESTIGATION

In this notebook you have been learning all these techniques to be able to manipulate your dataframe to your preference. We know how to clean and explore our data, but what questions or topics did you actually want to learn from the data? 

<b> * In groups of 2-4 people, investigate the dataframe in this notebook and pick a question/topic to answer. Using the techniques you learned today, show relationships and results that would support that question/topic. 
</b><br><br>
<i>If we have time</i> <b>each</b> group will present their investigations and why they are significant to the class.


In [624]:
df3 = df[df['LONGITUDE'].str.startswith('-122') & df['LATITUDE'].str.startswith('37')]
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5310 entries, 0 to 5616
Data columns (total 9 columns):
CASENO       5310 non-null int64
OFFENSE      5310 non-null object
EVENTDT      5310 non-null object
EVENTTM      5310 non-null object
EVENTDESC    5310 non-null object
D.O.W.       5310 non-null object
BLKADDR      5310 non-null object
LATITUDE     5310 non-null object
LONGITUDE    5310 non-null object
dtypes: int64(1), object(8)
memory usage: 414.8+ KB


In [625]:
df3['LONGITUDE'] = pd.to_numeric(df3['LONGITUDE'])
df3['LATITUDE'] = pd.to_numeric(df3['LATITUDE'])


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [626]:
import numpy
df3['LATITUDE'] = ((df3['LATITUDE'] + 0.005) * 100).astype(numpy.int64)/100

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [627]:
import numpy
df3['LONGITUDE'] = ((df3['LONGITUDE'] + 0.005) * 100).astype(numpy.int64)/100


/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [628]:
df3.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                         BLKADDR  LATITUDE  LONGITUDE  
0  Wednesday      ALLSTON WAY & SHATTUCK AVE     37.87    -122.26  
1     Friday   UNIVERSITY AVENUE & MILVIA ST     37.87    -122.26  
2   Saturday               2500 SHATTUCK AVE     37.86    -122.26  
3     Sunday    DURANT AVENUE & ELLSWORTH ST     37.87    -122.25  
4   Saturday  MILVIA STREET & UNIVERSITY AVE     37.87    -122.26

In [629]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5310 entries, 0 to 5616
Data columns (total 9 columns):
CASENO       5310 non-null int64
OFFENSE      5310 non-null object
EVENTDT      5310 non-null object
EVENTTM      5310 non-null object
EVENTDESC    5310 non-null object
D.O.W.       5310 non-null object
BLKADDR      5310 non-null object
LATITUDE     5310 non-null float64
LONGITUDE    5310 non-null float64
dtypes: float64(2), int64(1), object(6)
memory usage: 414.8+ KB


In [630]:
def locate(f):
    return (str(f['LATITUDE']) + " " + str(f['LONGITUDE']))
df3['LOCATION'] = df3.apply(locate,axis=1)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [631]:
df3.head()

CASENO                   OFFENSE     EVENTDT EVENTTM           EVENTDESC  \
0  17034187             BURGLARY AUTO  06/14/2017   15:15  BURGLARY - VEHICLE   
1  17052235                GUN/WEAPON  09/01/2017   22:56     WEAPONS OFFENSE   
2  17091126  THEFT MISD. (UNDER $950)  06/10/2017   10:45             LARCENY   
3  17018444             BURGLARY AUTO  04/02/2017   19:30  BURGLARY - VEHICLE   
4  17033328                 NARCOTICS  06/10/2017   14:30      DRUG VIOLATION   

      D.O.W.                         BLKADDR  LATITUDE  LONGITUDE  \
0  Wednesday      ALLSTON WAY & SHATTUCK AVE     37.87    -122.26   
1     Friday   UNIVERSITY AVENUE & MILVIA ST     37.87    -122.26   
2   Saturday               2500 SHATTUCK AVE     37.86    -122.26   
3     Sunday    DURANT AVENUE & ELLSWORTH ST     37.87    -122.25   
4   Saturday  MILVIA STREET & UNIVERSITY AVE     37.87    -122.26   

        LOCATION  
0  37.87 -122.26  
1  37.87 -122.26  
2  37.86 -122.26  
3  37.87 -122.25  
4  37.87 -122.26

In [632]:
df3.LOCATION.unique()

array(['37.87 -122.26', '37.86 -122.26', '37.87 -122.25', '37.88 -122.26',
       '37.88 -122.29', '37.86 -122.28', '37.86 -122.29', '37.87 -122.24',
       '37.82 -122.27', '37.89 -122.26', '37.89 -122.3', '37.86 -122.25',
       '37.85 -122.27', '37.86 -122.24', '37.87 -122.27', '37.86 -122.27',
       '37.87 -122.29', '37.88 -122.3', '37.88 -122.28', '37.87 -122.28',
       '37.85 -122.28', '37.88 -122.27', '37.85 -122.26', '37.88 -122.25',
       '37.85 -122.24', '37.89 -122.25', '37.86 -122.31', '37.89 -122.27',
       '37.85 -122.23', '37.86 -122.23', '37.87 -122.3', '37.9 -122.25',
       '37.85 -122.25', '37.87 -122.31', '37.9 -122.26', '37.89 -122.24',
       '37.9 -122.27', '37.9 -122.28', '37.88 -122.24', '37.89 -122.28',
       '37.85 -122.29', '37.87 -122.23', '37.91 -122.27', '37.86 -122.22'], dtype=object)

In [655]:
df4 = df3
df5=df4.LOCATION.value_counts().reset_index()
df5.rename(columns ={'index':'LOCATION', 'LOCATION':'OCCURRENCE'}, inplace=True)
df5.head()

LOCATION  OCCURRENCE
0  37.87 -122.26         891
1  37.87 -122.25         462
2  37.86 -122.26         339
3  37.87 -122.29         307
4  37.86 -122.25         291

In [656]:
df5 = df5.drop(df5.index[0])
df5.head()

LOCATION  OCCURRENCE
1  37.87 -122.25         462
2  37.86 -122.26         339
3  37.87 -122.29         307
4  37.86 -122.25         291
5  37.87 -122.28         283

In [658]:
df6 = df3[df3.LOCATION.isin(df5[df5['OCCURRENCE'] == df5['OCCURRENCE'].max()]['LOCATION'])]
len(df6.index)

462

In [659]:
#df5=df4.LOCATION.value_counts().reset_index()
df7 = df6['D.O.W.'].value_counts().reset_index()
df7.rename(columns ={'index':'DOW', 'D.O.W.':'OCCURRENCE'}, inplace=True)
df7.head()

DOW  OCCURRENCE
0   Saturday          77
1   Thursday          76
2  Wednesday          75
3     Monday          62
4     Friday          60

In [660]:
df7[df7['OCCURRENCE'] == df7['OCCURRENCE'].max()]['DOW']
df8 = df6[df6['D.O.W.'].isin(df7[df7['OCCURRENCE'] == df7['OCCURRENCE'].max()].DOW)]
len(df8.index)

77

In [661]:
df8

CASENO                   OFFENSE     EVENTDT EVENTTM  \
68    17021019             BURGLARY AUTO  04/15/2017   16:00   
77    17042833      ASSAULT/BATTERY FEL.  07/22/2017   23:30   
171   17045779               DISTURBANCE  08/05/2017   19:11   
197   17090755             BURGLARY AUTO  04/15/2017   23:00   
222   17091120  THEFT MISD. (UNDER $950)  06/10/2017   18:00   
323   17091471  THEFT MISD. (UNDER $950)  07/29/2017   12:30   
326   17050663     ASSAULT/BATTERY MISD.  08/26/2017   16:46   
327   17034897               DISTURBANCE  06/17/2017   17:02   
376   17021062              2ND RESPONSE  04/15/2017   23:28   
425   17031666      ASSAULT/BATTERY FEL.  06/03/2017   08:02   
434   17091455                 VANDALISM  07/22/2017   00:10   
471   17020995               DISTURBANCE  04/15/2017   18:59   
494   17041937       BURGLARY COMMERCIAL  07/15/2017   16:00   
562   17090762             BURGLARY AUTO  04/15/2017   17:10   
668   17053310               DISTURBANCE  09/02/2017   12:00   
686   17052514                   ROBBERY  09/02/2017   23:06   
690   17033347  THEFT MISD. (UNDER $950)  06/10/2017   17:16   
750   17091474             BURGLARY AUTO  07/29/2017   23:00   
784   17024312            VEHICLE STOLEN  04/29/2017   22:00   
935   17091859                 VANDALISM  09/16/2017   08:30   
1064  17054399             BURGLARY AUTO  09/09/2017   23:00   
1133  17044085  THEFT MISD. (UNDER $950)  07/29/2017   10:07   
1219  17041043             MISSING ADULT  07/15/2017   01:30   
1317  17091003  THEFT MISD. (UNDER $950)  05/20/2017   08:10   
1353  17091601             BURGLARY AUTO  08/12/2017   13:00   
1404  17052454                 NARCOTICS  09/02/2017   19:20   
1445  17091266  THEFT FELONY (OVER $950)  07/01/2017   03:00   
1634  17034835  THEFT MISD. (UNDER $950)  06/17/2017   13:25   
1648  17036425             BURGLARY AUTO  06/24/2017   10:15   
1726  17025628               DISTURBANCE  05/06/2017   19:47   
...        ...                       ...         ...     ...   
3136  17020871  THEFT FELONY (OVER $950)  04/15/2017   01:40   
3178  17019828     ASSAULT/BATTERY MISD.  04/08/2017   01:00   
3188  17090668             BURGLARY AUTO  04/01/2017   16:00   
3263  17031760  THEFT MISD. (UNDER $950)  06/03/2017   17:15   
3351  17030103           ALCOHOL OFFENSE  05/27/2017   02:02   
3513  17090728  THEFT MISD. (UNDER $950)  04/08/2017   15:05   
3526  17028659             BURGLARY AUTO  05/20/2017   17:30   
3579  17041044      ASSAULT/BATTERY FEL.  07/15/2017   02:42   
3638  17091487             BURGLARY AUTO  07/22/2017   16:45   
3652  17047296  THEFT MISD. (UNDER $950)  08/12/2017   14:50   
3719  17052278              2ND RESPONSE  09/02/2017   00:39   
4094  17020921             BURGLARY AUTO  04/15/2017   11:07   
4241  17047294               DISTURBANCE  08/12/2017   14:51   
4281  17022473                 NARCOTICS  04/22/2017   21:05   
4399  17091540             BURGLARY AUTO  07/29/2017   18:05   
4430  17020983  THEFT MISD. (UNDER $950)  04/15/2017   17:22   
4491  17021546                 NARCOTICS  04/15/2017   18:00   
4503  17091036  THEFT MISD. (UNDER $950)  05/20/2017   10:00   
4521  17018066  THEFT MISD. (UNDER $950)  04/01/2017   03:12   
4638  17039471               BRANDISHING  07/08/2017   08:42   
4691  17091690             BURGLARY AUTO  08/26/2017   11:40   
4786  17028669             BURGLARY AUTO  05/20/2017   13:30   
4789  17047333  THEFT MISD. (UNDER $950)  08/12/2017   17:23   
4829  17027184             BURGLARY AUTO  05/13/2017   15:00   
5257  17019484              2ND RESPONSE  04/08/2017   00:38   
5294  17035039       BURGLARY COMMERCIAL  06/17/2017   18:00   
5304  17054726  THEFT MISD. (UNDER $950)  09/09/2017   23:00   
5429  17091153             BURGLARY AUTO  06/17/2017   00:05   
5459  17050513      ASSAULT/BATTERY FEL.  08/26/2017   02:18   
5524  17091747                 VANDALISM  09/02/2017   02:00   

                  EVENTDESC    D.O.W.  

In [662]:
def group_by_4_hour(x):
    hour, _ = x['EVENTTM'].split(':')
    if int(hour) < 4 : return ('0:00 - 4:00')
    if int(hour) < 8 : return ('4:00 - 8:00')
    if int(hour) < 12 : return ('8:00 - 12:00')
    if int(hour) < 16 : return ('12:00 - 16:00')
    if int(hour) < 20 : return ('16:00 - 20:00')
    return ('20:00 - 24:00')
#period = [[i for i in ] for j in x['EVENTDT']]
df8['PERIOD'] = df8.apply(group_by_4_hour, axis=1)

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.


In [663]:
df8.head()

CASENO                   OFFENSE     EVENTDT EVENTTM  \
68   17021019             BURGLARY AUTO  04/15/2017   16:00   
77   17042833      ASSAULT/BATTERY FEL.  07/22/2017   23:30   
171  17045779               DISTURBANCE  08/05/2017   19:11   
197  17090755             BURGLARY AUTO  04/15/2017   23:00   
222  17091120  THEFT MISD. (UNDER $950)  06/10/2017   18:00   

              EVENTDESC    D.O.W.                      BLKADDR  LATITUDE  \
68   BURGLARY - VEHICLE  Saturday                2400 HASTE ST     37.87   
77              ASSAULT  Saturday  CHANNING WAY & ELLSWORTH ST     37.87   
171  DISORDERLY CONDUCT  Saturday           2300 TELEGRAPH AVE     37.87   
197  BURGLARY - VEHICLE  Saturday                2362 BANCROFT     37.87   
222             LARCENY  Saturday            2500 BENVENUE AVE     37.87   

     LONGITUDE       LOCATION         PERIOD  
68     -122.25  37.87 -122.25  16:00 - 20:00  
77     -122.25  37.87 -122.25  20:00 - 24:00  
171    -122.25  37.87 -122.25  16:00 - 20:00  
197    -122.25  37.87 -122.25  20:00 - 24:00  
222    -122.25  37.87 -122.25  16:00 - 20:00

In [664]:
df9 = df8.groupby(['PERIOD']).size().reset_index()
df9.rename(columns={0:'OCCURRENCE'}, inplace=True)
df9.head()

PERIOD  OCCURRENCE
0    0:00 - 4:00          20
1  12:00 - 16:00          11
2  16:00 - 20:00          23
3  20:00 - 24:00          14
4   8:00 - 12:00           9

In [665]:
df9[df9['OCCURRENCE'] == df9['OCCURRENCE'].max()]['OCCURRENCE']
#df7[df7['OCCURRENCE'] == df7['OCCURRENCE'].max()]['DOW']
#df8 = df6[df6['D.O.W.'].isin(df7[df7['OCCURRENCE'] == df7['OCCURRENCE'].max()].DOW)]
df10 = df8[df8['PERIOD'].isin(df9[df9['OCCURRENCE'] == df9['OCCURRENCE'].max()].PERIOD)]
len(df10.index)

23

In [666]:
df10.head()

CASENO                   OFFENSE     EVENTDT EVENTTM  \
68   17021019             BURGLARY AUTO  04/15/2017   16:00   
171  17045779               DISTURBANCE  08/05/2017   19:11   
222  17091120  THEFT MISD. (UNDER $950)  06/10/2017   18:00   
326  17050663     ASSAULT/BATTERY MISD.  08/26/2017   16:46   
327  17034897               DISTURBANCE  06/17/2017   17:02   

              EVENTDESC    D.O.W.             BLKADDR  LATITUDE  LONGITUDE  \
68   BURGLARY - VEHICLE  Saturday       2400 HASTE ST     37.87    -122.25   
171  DISORDERLY CONDUCT  Saturday  2300 TELEGRAPH AVE     37.87    -122.25   
222             LARCENY  Saturday   2500 BENVENUE AVE     37.87    -122.25   
326             ASSAULT  Saturday        2400 DANA ST     37.87    -122.25   
327  DISORDERLY CONDUCT  Saturday  2500 HILLEGASS AVE     37.87    -122.25   

          LOCATION         PERIOD  
68   37.87 -122.25  16:00 - 20:00  
171  37.87 -122.25  16:00 - 20:00  
222  37.87 -122.25  16:00 - 20:00  
326  37.87 -122.25  16:00 - 20:00  
327  37.87 -122.25  16:00 - 20:00

In [667]:
df10.groupby("EVENTDESC").OFFENSE.value_counts().to_frame()

OFFENSE
EVENTDESC             OFFENSE                          
ASSAULT               ASSAULT/BATTERY MISD.           1
BURGLARY - COMMERCIAL BURGLARY COMMERCIAL             2
BURGLARY - VEHICLE    BURGLARY AUTO                   6
DISORDERLY CONDUCT    DISTURBANCE                     4
DRUG VIOLATION        NARCOTICS                       2
FRAUD                 IDENTITY THEFT                  1
LARCENY               THEFT MISD. (UNDER $950)        6
SEX CRIME             SEXUAL ASSAULT MISD.            1